pycbc/events/coinc.py

```diff
@@ -245,92 +249,106 @@ def time_coincidence(t1, t2, window, slide_step=0):
+        # at the 2nd iteration, we are left with triggers for which both fixed
+        #  and pivot are coincident with ifo1
+        for ifo2 in ids:
+            logging.info('added ifo %s, testing against %s' % (ifo1, ifo2))
             w = win(ifo1, ifo2)
-            left = numpy.searchsorted(time, ctimes[ifo2] - w)
-            right = numpy.searchsorted(time, ctimes[ifo2] + w)
-
-            # remove elements that will not form a coinc
-            # There is only at most one trigger for an existing coinc
-            # (assumes triggers spaced > slide step)
+            left = numpy.searchsorted(time1, ctimes[ifo2] - w)
+            right = numpy.searchsorted(time1, ctimes[ifo2] + w)
+            # any times within time1 coincident with the time in ifo2 have
+            # indices between left and right
+            # nz indexes into times in ifo2 which have coincidences with ifo1
+            # times
             nz = (right - left).nonzero()
+            if len(nz[0]) and (right - left).max() > 1:
+                # we expect at most one coincident time in ifo1, assuming
+                # trigger spacing in ifo1 > time window
+                rlmax = (right - left).max()
+                where = right - left == rlmax
+                print([float(ti) for ti in time1[left[where][0]:right[where][0]]])
+                raise ValueError('Triggers in %s are closer than coincidence '
+                                 'window. This should not happen!' % ifo1)
+            # identify indices of times in ifo1 that form coincs with ifo2
+            # due to check above, at most one trigger in ifo1 per trigger in ifo2
             dep_ids = left[nz]
-
-            # The property that only one trigger can be within the window is ensured
-            # by the peak finding algorithm we use for each template.
-            # If that is modifed, this function may need to be
-            # extended.
-            if len(left) > 0 and (right - left).max() > 1:
-                raise ValueError('Somehow triggers are closer than time-delay window')
-
+            # slide is array of slide ids attached to pivot ifo
             slide = slide[nz]
-            for ifo in ctimes:
+
+            for ifo in ctimes:  # cycle over fixed and pivot
+                # (& previous additional ifos?)
+                # reduce times and IDs to just those forming a coinc with ifo1
                 ctimes[ifo] = ctimes[ifo][nz]
                 ids[ifo] = ids[ifo][nz]
 
-        # Add this detector now to the cumulative set and proceed to the next
-        # ifo coincidence test
-        ids[ifo1] = sort[dep_ids]
-        ctimes[ifo1] = otime[ids[ifo1]]
+        # undo time sorting on indices of ifo1 triggers
+        ids[ifo1] = tsort[dep_ids]
+        # FIXME do we want to do this?? add ifo1 times to ctimes which will be
+        # further reduced if more than one additional detector is tested
+        #ctimes[ifo1] = otime[ids[ifo1]]
```

Comment:
This should not be commented out. It is required if you have 4+ detectors that you check against all combinations.

Reply:
Can you please describe the exact desired behaviour with 4+ detectors.  What tests should be passed by the resulting coinc triggers?  At the moment it seems to be that the 4th detector's triggers are tested only against the 'pivot' and 'fixed' ones

Reply:
It's really the same behavior as for 2+ detectors. Every ifo must be coincident. You need to add any new detectors to this dict so that when another detector is checked for coincidences the past ones are *all* included. Before your patch this would be true, afterwards, it will just break I think, since it will still see an ifo in the ids dict but not have the coresponding ifo times.

Reply:
The loop that checks a new extra ifo against existing ones is `for ifo2 in ids:` however the previously added extra ifos are not added to the 'ids' dictionary, just to 'ctimes'.  So that made it unclear what was supposed to be happening

Reply:
hmm, not sure why I thought that if the code has `ids[ifo1] = sort[dep_ids]`

Reply:
see new diff.